lib/Doctrine/DBAL/Driver/PDOConnection.php

```diff
@@ -130,4 +168,47 @@ public function requiresQueryForServerVersion()
     {
         return false;
     }
+
+    public function trackLastInsertId()
+    {
+        if (! $this->supportsTrackingLastInsertId()) {
+            return;
+        }
+
+        // We need to avoid unnecessary exception generation for drivers not supporting this feature,
+        // by temporarily disabling exception mode.
+        $this->setAttribute(\PDO::ATTR_ERRMODE, \PDO::ERRMODE_SILENT);
+
+        $lastInsertId = parent::lastInsertId();
+
+        // Reactivate exception mode.
+        $this->setAttribute(\PDO::ATTR_ERRMODE, \PDO::ERRMODE_EXCEPTION);
+
+        if (null === $lastInsertId) {
+            // In case this driver implementation does not support this feature
+            // or an error occurred while retrieving the last insert ID, simply return the last tracked insert ID.
+            return;
+        }
+
+        // The last insert ID is reset to "0" in certain situations by some implementations,
+        // therefore we keep the previously set insert ID locally.
+        if ('0' !== $lastInsertId) {
+            $this->lastInsertId = $lastInsertId;
+        }
+    }
+
+    /**
+     * Returns whether this PDO driver is able to fetch the last insert ID for identity columns
+     * after each executed statement(even if it might not have created an ID)
+     * without influencing connection state and transaction state.
+     *
+     * Drivers like "pdo_pgsql" for example cannot safely try to retrieve the last insert ID without causing an error
+     * that might cause active transactions to fail.
+     *
+     * @return bool
+     */
+    protected function supportsTrackingLastInsertId()
```

Comment:
This can most probably be a constant or a protected property. Having it as a method seems excessive, and can lead to (although minor) overhead